lib/spack/spack/schema/environment.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+"""Schema for environment modifications. Meant for inclusion in other
+schemas.
+"""
+
+array_of_strings_or_num = {
+    'type': 'array', 'default': [], 'items':
+        {'anyOf': [{'type': 'string'}, {'type': 'number'}]}
+}
+
+dictionary_of_strings_or_num = {
+    'type': 'object', 'patternProperties':
+        {r'\w[\w-]*': {'anyOf': [{'type': 'string'}, {'type': 'number'}]}}
+}
+
+definition = {
+    'anyOf': [
+        {
+            'type': 'object',
+            'default': {},
+            'additionalProperties': False,
+            'properties': {
+                'set': dictionary_of_strings_or_num,
+                'unset': array_of_strings_or_num,
+                'prepend_path': dictionary_of_strings_or_num,
+                'append_path': dictionary_of_strings_or_num
+            }
+        }, {
+            'type': 'object',
+            'default': {},
+            'additionalProperties': False,
+            'properties': {
+                'set': dictionary_of_strings_or_num,
+                'unset': array_of_strings_or_num,
+                'prepend-path': dictionary_of_strings_or_num,
+                'append-path': dictionary_of_strings_or_num
+            }
+        }
```

Comment:
Note to reviewers: we can get rid of some boilerplate if we drop backward compatibility on this naming inconsistency. If we do so this will affect people that currently have `append-path` or `prepend-path` in their `compilers.yaml`.

Reply:
I think we should make that change. The number of people using it in `compilers.yaml` is likely low, and in the long run it's better to make the code clean than to support the old naming scheme.

Reply:
Completely agree with that!